nixos/modules/services/monitoring/grafana.nix

```diff
@@ -635,7 +637,7 @@ in {
     systemd.services.grafana = {
       description = "Grafana Service Daemon";
       wantedBy = ["multi-user.target"];
-      after = ["networking.target"];
+      after = ["networking.target"] ++ lib.optional usePostgresql "postgresql.service" ++ lib.optional useMysql "mysql.service";
```

Comment:
I don't know the systemd behavior here from memory, but have you checked what'll happen if `postgresql.service` doesn't exist (i.e. when using postgres for grafana, but connecting to a database instance on a different machine)?

Reply:
It would still work - see https://github.com/NixOS/nixpkgs/commit/9b970d07f384f31737cb3d598913aba16dbc59bb#diff-c4e76be6d8d687a02103da358697dd78a98738f3a1bc20c52b3a00fbccf2e6a9 for an example.

That being said... see my above comment about checking the host as well to confirm it is a local connection, because the reason you mention.

Reply:
full reference is here https://www.freedesktop.org/software/systemd/man/systemd.unit.html

The interesting part is here
```
Note that those settings are independent of and orthogonal to the requirement dependencies as configured by Requires=, Wants=, Requisite=, or BindsTo=. It is a common pattern to include a unit name in both the After= and Wants= options, in which case the unit listed will be started before the unit that is configured with these options.
```
As I understand it, we would need to specify it in after but also in either Wants or Requires

if we can determine if it's a local connection then a specification in Requires makes sense. If we can't then a specification in Wants should be more tolerant.

Reply:
I think `Wants` is not correct here.

In the reference of the `Wants` option it says
```
Units listed in this option will be started if the configuring unit is. However, if the listed units fail to start or cannot be added to the transaction, this has no impact on the validity of the transaction as a whole, and this unit will still be started. This is the recommended way to hook the start-up of one unit to the start-up of another unit.
```

As I understand it, if we specify `Wants=postgresql.service` in the grafana unit, the grafana unit would try to start the postgresql service  regardless of whether it's disabled or not. When just using `After`, the grafana unit will be started after postgresql if the postgresql service is enabled.

`services.gitea` also only uses `Wants`: https://github.com/NixOS/nixpkgs/blob/99d8d553da44fb065f9cb8bb54e2e93b85417fab/nixos/modules/services/misc/gitea.nix#L459

Reply:
The only reason to use wants would be to allow for the fact that the postgres connection might not be on local but remote. Putting Requires here would not start the grafana instance for people using a remote postgres connection.

So in my opinion since having a `useLocalPostgresql` seems more work, we wouldn't be able to use Require here. Let me know if that doesn't make sense.

Reply:
I made a mistake in my comment: I said "services.gitea also only uses Wants", but meant "services.gitea also only uses After".

---

It's correct that we can't use `Require` without a check if the database runs on the same host. As far as I understood After/Wants/Require we only need After and not Wants or Require. I think that both "Wants" and "Require" are not right for our use case, because we don't want the grafana service to influence the postgresql service (see my last comment in this thread).

Reply:
I agree with you.